realm/realm-library/src/objectServer/java/io/realm/PermissionManager.java

```diff
@@ -0,0 +1,388 @@
+            RealmLog.error("PermissionManager was not correctly closed before being finalized.");
+            close();
+        }
+    }
+
+    // Creates the URL to the permission/management Realm based on the authentication URL.
+    private static String getRealmUrl(RealmType type, URL authUrl) {
+        String scheme = "realm";
+        if (authUrl.getProtocol().equalsIgnoreCase("https")) {
+            scheme = "realms";
+        }
+        try {
+            return new URI(scheme, authUrl.getUserInfo(), authUrl.getHost(), authUrl.getPort(),
+                    "/~/" + type.getName(), null, null).toString();
+        } catch (URISyntaxException e) {
+            throw new IllegalArgumentException("Could not create URL to the " + type + " + Realm", e);
+        }
+    }
+
+    // Task responsible for loading the Permissions result and returning it to the user.
+    // The Permission result is not considered available until the query has complated.
+    private class PermissionsAsyncTask extends PermissionManagerAsyncTask {
+
+        private final Callback<RealmResults<Permission>> callback;
+        private RealmResults<Permission> loadingPermissons; // Prevent permissions from being GC'ed until fully loaded.
+
+        PermissionsAsyncTask(Callback<RealmResults<Permission>> callback) {
+            this.callback = callback;
+        }
+
+        @Override
+        public void run() {
+            if (isCancelled()) { return; }
+            if (permissions != null || isClosed()) {
+                notifyCallbackWithResult();
+            } else {
+                synchronized (errorLock) {
+                    if (permissionRealmError != null) {
+                        callback.onError(permissionRealmError);
+                    } else {
+                        loadingPermissons = permissionRealm.where(Permission.class).findAllAsync();
+                        loadingPermissons.addChangeListener(new RealmChangeListener<RealmResults<Permission>>() {
+                            @Override
+                            public void onChange(RealmResults<Permission> loadedPermissions) {
+                                loadingPermissons.removeChangeListener(this);
+                                if (isCancelled()) { return; }
+                                if (permissions == null) {
+                                    permissions = loadedPermissions;
+                                }
+                                notifyCallbackWithResult();
+                            }
+                        });
+                    }
+                }
+            }
+        }
+
+        private void notifyCallbackWithResult() {
+            synchronized (errorLock) {
+                if (isClosed()) {
+                    IllegalStateException reason = new IllegalStateException("PermissionManager has been closed.");
+                    callback.onError(new ObjectServerError(ErrorCode.UNKNOWN, reason));
```

Comment:
same...  If the callback is called with the lock, that ought to be part of the documented contract.